near-sdk/src/types/public_key.rs

```diff
@@ -0,0 +1,162 @@
+use serde::{Deserialize, Serialize};
+use std::convert::{TryFrom, TryInto};
+
+/// PublicKey curve
+#[derive(Serialize, Deserialize)]
+pub enum CurveType {
+    ED25519 = 0,
+    SECP256K1 = 1,
+}
+
+impl TryFrom<String> for CurveType {
+    type Error = Box<dyn std::error::Error>;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        match value.to_lowercase().as_str() {
+            "ed25519" => Ok(CurveType::ED25519),
+            "secp256k1" => Ok(CurveType::SECP256K1),
+            _ => Err("Unknown curve kind".into()),
+        }
+    }
+}
+
+/// Public key in a binary format with string serialization.
+/// e.g. `ed25519:3tysLvy7KGoE8pznUgXvSHa4vYyGvrDZFcT8jgb8PEQ6`
+#[derive(Clone, PartialEq, PartialOrd, Ord, Eq)]
+pub struct StrPublicKey(pub Vec<u8>);
+
+impl StrPublicKey {
+    fn split_key_type_data(value: &str) -> Result<(CurveType, &str), Box<dyn std::error::Error>> {
+        if let Some(idx) = value.find(':') {
+            let (prefix, key_data) = value.split_at(idx);
+            Ok((CurveType::try_from(prefix.to_string())?, &key_data[1..]))
+        } else {
+            // If there is no Default is ED25519.
+            Ok((CurveType::ED25519, value))
+        }
+    }
+}
+
+impl From<StrPublicKey> for Vec<u8> {
+    fn from(v: StrPublicKey) -> Vec<u8> {
+        v.0
+    }
+}
+
+impl TryFrom<Vec<u8>> for StrPublicKey {
+    type Error = Box<dyn std::error::Error>;
+
+    fn try_from(v: Vec<u8>) -> Result<Self, Self::Error> {
+        match v.len() {
+            33 if v[0] == 0 => Ok(Self(v)),
+            65 if v[0] == 1 => Ok(Self(v)),
+            _ => Err("Invalid public key".into()),
+        }
+    }
+}
+
+impl serde::Serialize for StrPublicKey {
+    fn serialize<S>(
+        &self,
+        serializer: S,
+    ) -> Result<<S as serde::Serializer>::Ok, <S as serde::Serializer>::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_str(&String::from(self))
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for StrPublicKey {
+    fn deserialize<D>(deserializer: D) -> Result<Self, <D as serde::Deserializer<'de>>::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let s = <String as serde::Deserialize>::deserialize(deserializer)?;
+        s.try_into()
+            .map_err(|err: Box<dyn std::error::Error>| serde::de::Error::custom(err.to_string()))
+    }
+}
+
+impl From<&StrPublicKey> for String {
+    fn from(str_public_key: &StrPublicKey) -> Self {
+        match str_public_key.0[0] {
+            0 => "ed25519:".to_string() + &bs58::encode(&str_public_key.0[1..]).into_string(),
```

Comment:
We are considering switching from b58 encoding in the future https://github.com/nearprotocol/nearcore/issues/2100
Let's not add it here since we won't be able to remove it from the compiled smart contracts in the future. Also, `bs58` is not a very popular library and I would suggests us to be conservative when deciding to add dependencies to `near-sdk`. I don't see a good reason why we won't `PublicKey` as a regular enum with `Vec<u8>` stringified as hex or base64.

Reply:
I don't think we'll ever switch from bs58 for public keys, but it would make sense to rename `StrPublicKey` to `PublicKeyBase58` to make it clear and forward compatible.
The format right now is used by our tools and RPCs, so it's a standard right now.